pkg/controller/apiextensions/publication/watch.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package publication
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+)
+
+type adder interface {
+	Add(item interface{})
+}
+
+// EnqueueRequestForRequirement enqueues a reconcile.Request for the
+// NamespacedName of a RequirementReferencer's RequirementReference.
+type EnqueueRequestForRequirement struct{}
```

Comment:
Why is this necessary?

Reply:
It's analogous to the existing `resource.EnqueueRequestForClaim` from crossplane-runtime. It allows the requirement controller to watch composite resources of the kind that it is interested in (e.g. the `MySQLRequirement` controller can watch for `MySQL`). This allows the requirement controller to avoid speculatively requeuing reconciles when it's waiting for the composite resource to do something; it can instead just return `Requeue: false` and wait for a watch event.

We can't use the upstream `EnqueueRequestForOwner{}` here, because the (namespaced) requirement is not the owner of the (cluster scoped) composite.